### ISSUE TITLE
Build hangs if the target C++ compiler errors out

### ISSUE THREAD

**Issue by cmnrd** [icyphy/lingua-franca#130]
There is a strange issue with compiling for the C++ target. When make is called by lfc and an error occurs during compilation, the make process does not terminate (as happened in the [latest build](https://travis-ci.com/github/icyphy/lingua-franca/builds/156414521)) . This makes lfc wait forever. Killing make manually terminates the process and also lets lfc terminate and print the error messages from make. Oddly, this works fine when running make directly from the command line and it also works when compiling in Eclipse. Only compilation in lfc seems to be affected.

Does anyone have an idea what might cause this behavior?

**Comment by lhstrh** [comment]
Have you tried running `java -jar org.icyphy.linguafranca-1.0.0-SNAPSHOT-all.jar file.lf` to see whether the problem is in the jar or in `lfc`?

**Comment by lhstrh** [comment]
Also, is this the same issue as #85?

**Comment by cmnrd** [comment]
> Also, is this the same issue as #85?
I don't think so. In my case it got stuck at running make and never continued. There isn't any loop. Those two issues could have a common cause though.

> Have you tried running java -jar org.icyphy.linguafranca-1.0.0-SNAPSHOT-all.jar file.lf to see whether the problem is in the jar or in lfc?
No I didn't. Oddly I didn't encounter this error in a while now. I will wait and try this if I should encounter the problem again.

**Comment by cmnrd** [comment]
I found the source of this problem. Java writes the standard output and error streams of the process to a buffer with limited size. When the buffer is full, the process blocks and waits for the buffer to be emptied. In our case, the data was only read after the process terminates. Therfore, there was a deadlock.

I added a new method runSubprocess() which takes care of this. It spawns two threads and forwards the output and error streams to the system streams. Apart from changing the C++ generator to use the new method, I also took the liberty to modify the executeCommand() method whicht the other targets use.

BTW: I would also like to use the executeCommand() method for C++, but I need more control on the process builder than the method currently offers. Also, I really do not want to execute make a second time if it fails. I think the PATH problem could be solved more elegantly. Why not check for the given command first using `which <comman>` as is done for cmake in the C++ target?
